prediction_market_agent/agents/ofvchallenger_agent/deploy.py

```diff
@@ -1,21 +1,146 @@
+    "0x9D0260500ba7b068b5b0f4AfA9F8864eBc0B059a"
+)
+CHALLENGE_BOND = xdai_type(10)
+# We benchmarked OFV only against Olas market-creators.
+MARKET_CREATORS_TO_CHALLENGE: list[ChecksumAddress] = [
+    # Olas market-creator 0.
+    Web3.to_checksum_address("0x89c5cc945dd550bcffb72fe42bff002429f46fec"),
+    # Olas market-creator 1.
+    Web3.to_checksum_address("0xffc8029154ecd55abed15bd428ba596e7d23f557"),
+]
+
+
+class Challenge(BaseModel):
+    old_responses: list[RealityResponse]
+    new_resolution: Resolution
 
 
 class OFVChallengerAgent(DeployableAgent):
     def run(self, market_type: MarketType) -> None:
         if market_type != MarketType.OMEN:
             raise RuntimeError("Can challenge only Omen.")
 
-        self.challenge()
+        api_keys = APIKeys()
+        self.challenge(api_keys)
 
     @observe()
-    def challenge(self) -> None:
+    def challenge(self, api_keys: APIKeys) -> None:
         self.langfuse_update_current_trace(tags=[OFV_CHALLENGER_TAG])
-        APIKeys()
-        # TODO: Coming in the next PR.
+
+        # Claim the bonds as first thing, to have funds for the new challenges.
+        claim_all_bonds_on_reality(api_keys)
+
+        markets_open_for_answers = OmenSubgraphHandler().get_omen_binary_markets(
+            limit=None,
+            creator_in=MARKET_CREATORS_TO_CHALLENGE,
+            # We need markets already opened for answers.
+            opened_before=utcnow(),
+            # With a little bandwidth for the market to be finalized,
+            # so we have time for processing it without erroring out at the end.
+            finalized_after=utcnow() + timedelta(minutes=30),
+        )
+        logger.info(f"Found {len(markets_open_for_answers)} markets to challenge.")
+
+        for market in markets_open_for_answers:
+            self.challenge_market(market, api_keys)
+
+    @observe()
+    def challenge_market(
+        self,
+        market: OmenMarket,
+        api_keys: APIKeys,
+        web3: Web3 | None = None,
+    ) -> Challenge | None:
+        logger.info(f"Challenging market {market.url=}")
+
+        existing_responses = OmenSubgraphHandler().get_responses(
+            question_id=market.question.id
+        )
+        logger.info(
+            f"{market.url=}'s responses and bonds: {[(r.answer, r.bond_xdai) for r in existing_responses]}"
+        )
+
+        # Next bond needs to be at least double the previous one.
+        if any(
+            response.bond_xdai >= CHALLENGE_BOND / 2 for response in existing_responses
+        ):
+            logger.info(
+                f"Market {market.url=} already challenged with bond > {CHALLENGE_BOND} / 2. Skipping."
+            )
+            return None
+
+        # We don't plan to re-challenge markets already challenged by the challenger, should we?
+        if any(
+            response.user_checksummed == OFV_CHALLENGER_SAFE_ADDRESS
+            for response in existing_responses
+        ):
```

Comment:
This should already be covered by the previous check, right? Might be good to move it before it though, so you get more informative logging (this message if we've already challenged it vs other message if someone else has challenged it)

Reply:
Yeah, ideally it should be. Flipped them.